graphql/src/runner.rs

```diff
@@ -135,12 +141,21 @@ where
     ) -> Self {
         let logger = logger.new(o!("component" => "GraphQlRunner"));
         let result_size = Arc::new(ResultSizeMetrics::new(registry));
+        let mut graphql_validation_plan = ValidationPlan { rules: Vec::new() };
+
+        graphql_validation_plan.add_rule(Box::new(LoneAnonymousOperation {}));
+        graphql_validation_plan.add_rule(Box::new(FragmentsOnCompositeTypes {}));
+        graphql_validation_plan.add_rule(Box::new(OverlappingFieldsCanBeMerged {}));
+        graphql_validation_plan.add_rule(Box::new(KnownFragmentNamesRule {}));
+        graphql_validation_plan.add_rule(Box::new(NoUnusedFragments {}));
+
```

Comment:
It would be better IMHO to move this and the call to validating the plan below to `graphql::execution::Query` either as part of `query.validate_fields` [here](https://github.com/graphprotocol/graph-node/blob/master/graphql/src/execution/query.rs#L379) or as a new method `validate_query` on that struct.

Reply:
I think this will change the nature of `validate` phase to be Query-specific, no? (some validation rules are useful for mutations, and some for subscriptions, so the `validate` phase is for every kind of operation, not only Query...)

I agree that we should find a better place to run it, I was thinking about doing it in `node/src/manager/commands/query.rs` in-between `parse` and `execute`, but we don't have the  `schema::Document` on that point. 

Also, I guess we can remove the `validate_fields` at some point, since it will be executed as validation rule (KnownFields), so it will also run on a separate layer from `Query` ?

Reply:
> I think this will change the nature of `validate` phase to be Query-specific, no? (some validation rules are useful for mutations, and some for subscriptions, so the `validate` phase is for every kind of operation, not only Query...)

The `graphql::execution::query::Query` struct is actually used for both subscriptions and queries, so moving it there wouldn't bypass any logic. As for mutations, they are so far out for this part of the code that I don't think it's worth worrying about them.

> I agree that we should find a better place to run it, I was thinking about doing it in `node/src/manager/commands/query.rs` in-between `parse` and `execute`, but we don't have the `schema::Document` on that point.

The code in `node/src/manager` is only used by `graphman` commands, and not used in the `graph-node` service; putting validation there would have no effect for the service.

> Also, I guess we can remove the `validate_fields` at some point, since it will be executed as validation rule (KnownFields), so it will also run on a separate layer from `Query` ?

Yes, that's right; one important wrinkle is [this comment](https://github.com/graphprotocol/graph-node/blob/master/graphql/src/execution/query.rs#L178-L179): it should be a little clearer, but what that is talking about is mutually recursive fragments which can lead to infinite recursion if you just follow them. I haven't looked at the details with `graphql-tools`, but we have to either move any check that is following fragments after the call to `check_complexity` or make sure that `graphql-tools` checks that itself.

Reply:
I see, let me check, the spec has a rule for recursive fragments, so maybe we can add that as well, and it will simplify some of the checks `graph-node` needs to to today. 

Regarding `Query`, I see what you mean - I'll move it there :) thanks!

Reply:
Done ;) thanks @lutter